### ISSUE TITLE
Clip gradients for fixed point arithmetics

### ISSUE THREAD

**Issue by glencoe** [es-ude/elastic-ai.creator#264]
Gradients of quantization functions using STE should typically be clipped whenever the input exceeds the representable numeric range. Possible implementation

```python
class _QuantizeFunc(torch.autograd.Function):
    @staticmethod
    def forward(ctx: Any, *args: Any, **kwargs: Any) -> Any:
        x = args[0]
        scale = args[1]
        bits = args[2]
        x = scale * x
        upper_bound = 2. ** bits - 1
        lower_bound = -2. ** bits
        mask = torch.logical_or(x > upper_bound, x < lower_bound)
        derivative = torch.where(mask, 0., 1.)
        ctx.save_for_backward(derivative)
        x = torch.round(torch.clamp(x, lower_bound, upper_bound))
        return x

    @staticmethod
    def backward(ctx: Any, *grad_outputs: Any) -> Any:
        derivative,  = ctx.saved_tensors
        return derivative * grad_outputs[0], None, None
```

**Comment by julianhoever** [comment]
This is already implemented, because we clamp the inputs first in the FixedPointArithmetics qunatize function and then apply a rounding. The backward pass of the torch clamp function already clips the gradients in the same way as described in this issue.
